### ISSUE TITLE
Missing `DM_DEV_ARM_POLL_CMD` in older verity header (files)

### ISSUE THREAD

**Issue by flxo** [stratis-storage/devicemapper-rs#686]
The `DM_DEV_ARM_POLL_CMD` has been added in [Linux 4.13](https://elixir.bootlin.com/linux/v4.13/source/include/uapi/linux/dm-ioctl.h#L244). Older headers (including the header that come with the [cross](https://github.com/rust-embedded/cross) Docker image for `aarch64-unknown-linux-gnu` are lacking this enum entry:

```
   --> /cargo/git/checkouts/devicemapper-rs-fdb8684edc8b6c26/851215b/src/core/dm.rs:766:28
    |
766 |         self.do_ioctl(dmi::DM_DEV_ARM_POLL_CMD as u8, &mut hdr, None)?;
    |                            ^^^^^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `DM_DEV_REMOVE_CMD`
    | 
   ::: /target/aarch64-unknown-linux-gnu/debug/build/devicemapper-bf7482b0eef6eb87/out/bindings.rs:584:1
    |
584 | pub const DM_DEV_REMOVE_CMD: ::std::os::raw::c_uint = 4;
    | -------------------------------------------------------- similarly named constant `DM_DEV_REMOVE_CMD` defined here

error[E0425]: cannot find value `DM_DEV_ARM_POLL_CMD` in module `dmi`
   --> /cargo/git/checkouts/devicemapper-rs-fdb8684edc8b6c26/851215b/src/core/dm.rs:766:28
    |
766 |         self.do_ioctl(dmi::DM_DEV_ARM_POLL_CMD as u8, &mut hdr, None)?;
    |                            ^^^^^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `DM_DEV_REMOVE_CMD`
    | 
   ::: /target/debug/build/devicemapper-970039da098373e2/out/bindings.rs:584:1
    |
584 | pub const DM_DEV_REMOVE_CMD: ::std::os::raw::c_uint = 4;
    | -------------------------------------------------------- similarly named constant `DM_DEV_REMOVE_CMD` defined here

error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
error: could not compile `devicemapper`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

Unfortunately it's not possible to manually inject this when using `devicemapper-rs` by e.g setting `BINDGEN_EXTRA_CLANG_ARGS` or other hacks. I think in theory the availability of `DM_DEV_ARM_POLL_CMD` should be tied to the device mapper version. I didn't find a way to get devicemapper-rs built without any modification.

Also the ioctl command `DM_GET_TARGET_VERSION_CMD` has been added "recently" and probably would have the same issue if it would be used by this crate.

I would be for happy with a simple `#cfg(not(target(...)))` for `aarch64-unknown-linux-gnu` above `core::DM::arm_poll` to avoid a fork since I don't need the function.

Optionions?

**Comment by mulkieran** [comment]
We have been evolving techniques in our other Rust libraries for cross-platform and version compatibility. I think we can use or adapt those techniques to retrofit the compatibility here.

**Comment by mulkieran** [comment]
@jbaublitz I believe that techniques that you use in libcryptsetup could be adapted more or less unchanged to address this problem. I noticed that you had an issue in libcryptsetup, https://github.com/stratis-storage/libcryptsetup-rs/issues/169, that also might be relevant here. Let me know what you think, thx.

**Comment by jbaublitz** [comment]
Hi @mulkieran! I think the technique could absolutely be adopted here with the caveat that `DEV_ARM_POLL` is the only asynchronous way to listen for devicemapper events. @flxo If that's not something you need, then this seems like something worth doing! I think the best way to handle this would be to check the devicemapper version in a build script, but the drawback of this is that I think this means devicemapper would have to be loaded in the kernel to build this library. Perhaps that's not a big deal though given the prevalence of devicemapper usage in many Linux distros.

@mulkieran I'm still thinking about how to best address that issue that you linked because there does not seem to be any way to compare values using macro attributes which was my original plan. After doing one last pass over what my options are, I'm going to close that one out.

**Comment by mulkieran** [comment]
@jbaublitz I agree that knowing what to check to set the cfg directives is the judgement bit. I guess our other options are checking the kernel version and letting the user decide via specifying features when building?

**Comment by flxo** [comment]
Thanks for your thoughts.

Build time checks for the device mapper version doesn't work for cross compilations. Here you just need to know what you're doing.

A build script is something I really like to avoid because I enjoy plain `cargo ...` or `cross ...` commands.
The only thing I could thing about is to gather the target dm (or  kernel) version in the `devicemapper-rs`s build.rs and set features accordingly. Getting the target dm version could come from a env variable, the driver, headers... in a specific precedence with env variables being the highest. I think relying on the kernel version if probably easier for users of devicemapper-rs than figuring out the dm version of the kernel in use.

Env variables would work for me, since my cross compilations rely on Docker images. I need custom docker images anyway and adding a DM target (or kernel) version in my docker files doesn't hurt and allow adopters to at least check the version to match. By checking I mean: Set the DM version to the version on target and see if the build is still fine.

**Comment by mulkieran** [comment]
@flxo Ok! Btw, when @jbaublitz uses the phrase "build script" he means "build.rs". So you two are on the same page there.

**Comment by jbaublitz** [comment]
@flxo Here is the issue that I filed with bindgen about my preferred solution although I highly doubt this is supported right now. https://github.com/rust-lang/rust-bindgen/issues/2095

**Comment by mulkieran** [comment]
@flxo @jbaublitz Until we get a more principled system through bindgen would something as simple as an environment variable be ok, as in #692? In the current design, if the env variable is not specified then everything gets compiled in, but that can be changed, obviously.

**Comment by flxo** [comment]
> @flxo @jbaublitz Until we get a more principled system through bindgen would something as simple as an environment variable be ok, as in #692? In the current design, if the env variable is not specified then everything gets compiled in, but that can be changed, obviously.

Thanks for the patch! Setting the variable is not a burden until bindgen does the job for you.

**Comment by mulkieran** [comment]
@jbaublitz My first thought was features, if we leave it to the user to specify what they want[1]. My reasons for thinking an environment variable would be better were complicated. I've just found out something new which inclines me to features more. Here[2] is an interesting crate, ```conf_test```, which sets feature flags from build.rs using an autotools-like approach with ```rustc --cfg```. What that means is that we could just use features now, but if in future we wanted to automate the feature selection we could use something like ```conf_test``` to test the version and set the feature as appropriate using autotools-like methods.

My further thought is to wonder if a related approach would be preferable for libcryptsetup as well. Would you prefer to use features and set the feature flags in the build script based on the information obtained via the ```pkg_config``` crate using the ```rustc --cfg``` command?

@flxo Would you be just as happy if the conditional compilation were feature-based?

[1] https://github.com/stratis-storage/devicemapper-rs/issues/686#issuecomment-911903167
[2] https://crates.io/crates/conf_test

**Comment by jbaublitz** [comment]
@mulkieran I'm not sure that it's good to mix features in `libcryptsetup-rs` when we already automatically determine which config options to set. This would allow users to also set features, and I'm not sure what the interaction would or even should be in that case. Part of the reason for features here is that we don't have an automatic way to determine which features should be enabled without reimplementing logic from other projects.

**Comment by flxo** [comment]
@mulkieran Thanks. I'm happy with that. I can simply turn off the `default-features` since in that case I never need `ARM_POLL`.

**Comment by mulkieran** [comment]
Ok! features it is.

**Comment by mulkieran** [comment]
Some useful facts:
* ```dm-ioctl.h``` contains three version constants, what that means is that our low-level bindings know their own version.
* ```device_mapper/ioctl/libdm-iface.c``` keeps notes about when a new constant is introduced

We could do something like generate the low-level bindings, compile a small script that makes the version number available to the build script, and then take advantage of that by setting a cfg environment variable that is respected by the compiler.

**Comment by mulkieran** [comment]
Closed by #698
